### ISSUE TITLE
Moshi needs an @JsonIgnore annotation, so fields can be independently excluded from serialization

### ISSUE THREAD

**Issue by bpappin** [square/moshi#1301]
**tl;dr** Using _transient_ keyword to ignore fields causes side effects, and is impossible in a kotlin data class. Moshi needs a _@JsonIgnore_ annotation similar to Rooms _@Ignore_, or similar to the inverse functionality found with Gson _@Expose_, so that we can isolate fields Moshi should ignore from the rest of the code, and avoid fields marked _transient_ from being interpreted unexpectedly. 

In order to ignore a field in Moshi, the transient keyword is used.
However this causes side effects with other libraries that also interpret the transient keyword.

Kotlin example:
Consider a data class that can be loaded from a remote source, deserialized through moshi, but also has a local representation stored by Room:

```
@Entity(tableName = "my_data")
@JsonClass(generateAdapter = true)
data class MyData(
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "local_id") 
    @Json(name = "remote_id") 
    val id:String, 
   
    @ColumnInfo(name = "local_name")
    @Json(name = "remote_name")   
    val name:String, 

    @ColumnInfo(name = "local_url") 
    // should be ignored by Moshi in both directions
    var url:String? = "http://example.com"
)
```

In a Java class similar to that Kotlin data class, the "url" field would need a _transient_ qualifier, **which would also cause Room to ignore it**.

**All of these problems can be alleviated by including a different mechanism for ignoring fields, such as including a _@JsonIgnore_ annotation.**

**Comment by ZacSweers** [comment]
The right solution is not to mix your domain models. It can be a convenient shortcut, but when they incur confusing and asymmetric side effects you end up in situations like this. 

We have considered adding a transient option to `@Json` for computed properties, but wouldn't do it for the domain conflation scenario you've described.

**Comment by bpappin** [comment]
First of all, I disagree, 80% of this is about maintenance, and how a model is structures is not defined by Moshi. 

It's not for Moshi to determine the internal architecture of an application, be it a new greenfield build or some legacy project that someone needs to work with. 

**The fact of the matter is that Moshi should not be using "transient" to ignore fields.**

it's extremely lame that this FR was close without properly evaluating the issues involved.

**Comment by bpappin** [comment]
The fact is that sometimes properties need to be hidden from Moshi without using the transient keyword.

**Comment by JakeWharton** [comment]
It was evaluated when the library was designed years ago. I would be interested in any use case that does not involve mixing domains. But if it's solely that you want to persist your models directly to the DB and serialize to JSON we're comfortable recommending that you either split the models or use a different library.
